File: problem/boj/2839.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const target = +require("fs").readFileSync(filePath, "utf-8").toString().trim();

const impossible = -1;
const memo = Array.from({ length: target + 1 }, () => impossible);
function BFS(init) {
  const stack = [];
  stack.push({ leftover: init, depth: 0 });
  memo[0] = 0;
  while (stack.length) {
    const { leftover, depth } = stack.shift();
    for (let sugar of [3, 5]) {
      if (memo[leftover + sugar] === impossible) {
        memo[leftover + sugar] = depth + 1;
        stack.push({ leftover: leftover + sugar, depth: depth + 1 });
      }
    }
  }
  console.log(memo[target]);
}
BFS(0);

File: problem/boj/2839.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;


const int impossible = -1;



struct Node {
    int leftover;
    int depth;
};

void BFS(int init, int target) {
    queue<Node> q;
    vector<int> memo(target + 1, impossible);
    q.push({init, 0});
    memo[0] = 0;

    while (!q.empty()) {
        Node node = q.front();
        q.pop();

        for (int sugar : {3, 5}) {
            if (node.leftover + sugar <= target && memo[node.leftover + sugar] == impossible) {
                memo[node.leftover + sugar] = node.depth + 1;
                q.push({node.leftover + sugar, node.depth + 1});
            }
        }
    }

    cout << memo[target] << endl;
}

int main() {
    int n;
    cin >> n;
    BFS(0, n);
    return 0;
}
